diff --git a/src/daemon/client/client.js b/src/daemon/client/client.js
index bb399296f1299044c0a06d2bf098ae15e6281b6c..514cfe01d12cdee19ef0ae0ec448d2b21767597e 100644
--- a/src/daemon/client/client.js
+++ b/src/daemon/client/client.js
@@ -869,7 +869,7 @@ class DaemonClient {
         // An open promise isn't enough to keep the event loop
         // alive, so we set a timeout here and clear it when we hear
         // back
-        const keepAlive = setTimeout(() => { }, 10 * 60 * 1000);
+        const keepAlive = setTimeout(() => { }, 4 * 60 * 1000);
         return new Promise((resolve, reject) => {
             perf_hooks_1.performance.mark('sendMessageToDaemon-start');
             this.currentMessage = message;
diff --git a/src/project-graph/plugins/isolation/isolated-plugin.js b/src/project-graph/plugins/isolation/isolated-plugin.js
index 64150514a8d72011c8794fd295fd2236cc8882bd..94778d5cfc54442a1c0fb32873351bc4a9bafb6b 100644
--- a/src/project-graph/plugins/isolation/isolated-plugin.js
+++ b/src/project-graph/plugins/isolation/isolated-plugin.js
@@ -12,7 +12,7 @@ const resolve_plugin_1 = require("../resolve-plugin");
 const messaging_1 = require("./messaging");
 const plugin_lifecycle_manager_1 = require("./plugin-lifecycle-manager");
 const PLUGIN_TIMEOUT_HINT_TEXT = 'As a last resort, you can set NX_PLUGIN_NO_TIMEOUTS=true to bypass this timeout.';
-const MINUTES = 10;
+const MINUTES = 3;
 const MAX_MESSAGE_WAIT = process.env.NX_PLUGIN_NO_TIMEOUTS === 'true'
     ? // Registering a timeout prevents the process from exiting
         // if the call to a plugin happens to be the only thing
