// nx@23.0.0-beta.12/dist/src/daemon/server/server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.openSockets = void 0;
exports.handleResult = handleResult;
exports.startServer = startServer;
const fs_1 = require("fs");
const net_1 = require("net");
const path_1 = require("path");
const v8_1 = require("v8");
const analytics_1 = require("../../analytics");
const file_hasher_1 = require("../../hasher/file-hasher");
const native_1 = require("../../native");
const consume_messages_from_socket_1 = require("../../utils/consume-messages-from-socket");
require("../../utils/perf-logging");
const versions_1 = require("../../utils/versions");
const workspace_context_1 = require("../../utils/workspace-context");
const workspace_root_1 = require("../../utils/workspace-root");
const nx_json_1 = require("../../config/nx-json");
const get_plugins_1 = require("../../project-graph/plugins/get-plugins");
const cache_1 = require("../cache");
const is_nx_version_mismatch_1 = require("../is-nx-version-mismatch");
const installed_nx_version_1 = require("../../utils/installed-nx-version");
const logger_1 = require("../logger");
const configure_ai_agents_1 = require("../message-types/configure-ai-agents");
const daemon_environment_1 = require("../client/daemon-environment");
const daemon_message_1 = require("../message-types/daemon-message");
const flush_sync_generator_changes_to_disk_1 = require("../message-types/flush-sync-generator-changes-to-disk");
const force_shutdown_1 = require("../message-types/force-shutdown");
const get_context_file_data_1 = require("../message-types/get-context-file-data");
const get_files_in_directory_1 = require("../message-types/get-files-in-directory");
const get_nx_workspace_files_1 = require("../message-types/get-nx-workspace-files");
const get_registered_sync_generators_1 = require("../message-types/get-registered-sync-generators");
const get_sync_generator_changes_1 = require("../message-types/get-sync-generator-changes");
const glob_1 = require("../message-types/glob");
const hash_glob_1 = require("../message-types/hash-glob");
const nx_console_1 = require("../message-types/nx-console");
const register_project_graph_listener_1 = require("../message-types/register-project-graph-listener");
const run_tasks_execution_hooks_1 = require("../message-types/run-tasks-execution-hooks");
const task_history_1 = require("../message-types/task-history");
const update_workspace_context_1 = require("../message-types/update-workspace-context");
const socket_utils_1 = require("../socket-utils");
const file_change_events_1 = require("./file-watching/file-change-events");
const route_workspace_changes_1 = require("./file-watching/route-workspace-changes");
const file_watcher_sockets_1 = require("./file-watching/file-watcher-sockets");
const handle_configure_ai_agents_1 = require("./handle-configure-ai-agents");
const handle_context_file_data_1 = require("./handle-context-file-data");
const handle_flush_sync_generator_changes_to_disk_1 = require("./handle-flush-sync-generator-changes-to-disk");
const handle_force_shutdown_1 = require("./handle-force-shutdown");
const handle_get_files_in_directory_1 = require("./handle-get-files-in-directory");
const handle_get_registered_sync_generators_1 = require("./handle-get-registered-sync-generators");
const handle_get_sync_generator_changes_1 = require("./handle-get-sync-generator-changes");
const handle_glob_1 = require("./handle-glob");
const handle_hash_glob_1 = require("./handle-hash-glob");
const handle_hash_tasks_1 = require("./handle-hash-tasks");
const handle_nx_console_1 = require("./handle-nx-console");
const handle_nx_workspace_files_1 = require("./handle-nx-workspace-files");
const handle_outputs_tracking_1 = require("./handle-outputs-tracking");
const handle_process_in_background_1 = require("./handle-process-in-background");
const handle_request_project_graph_1 = require("./handle-request-project-graph");
const handle_request_shutdown_1 = require("./handle-request-shutdown");
const handle_task_history_1 = require("./handle-task-history");
const handle_tasks_execution_hooks_1 = require("./handle-tasks-execution-hooks");
const handle_update_workspace_context_1 = require("./handle-update-workspace-context");
const outputs_tracking_1 = require("./outputs-tracking");
const project_graph_incremental_recomputation_1 = require("./project-graph-incremental-recomputation");
const project_graph_listener_sockets_1 = require("./project-graph-listener-sockets");
const shutdown_utils_1 = require("./shutdown-utils");
const sync_generators_1 = require("./sync-generators");
const watcher_1 = require("./watcher");
let workspaceWatcherError;
let outputsWatcherError;
global.NX_DAEMON = true;
process.env.NX_DAEMON_PROCESS = 'true';
let numberOfOpenConnections = 0;
exports.openSockets = new Set();
const server = (0, net_1.createServer)(async (socket) => {
    numberOfOpenConnections += 1;
    exports.openSockets.add(socket);
    logger_1.serverLogger.log(`Established a connection. Number of open connections: ${numberOfOpenConnections}`);
    (0, shutdown_utils_1.resetInactivityTimeout)(handleInactivityTimeout);
    socket.on('data', (0, consume_messages_from_socket_1.consumeMessagesFromSocket)(async (message) => {
        await handleMessage(socket, message);
    }));
    socket.on('error', (e) => {
        logger_1.serverLogger.log(`Socket error: ${e.message}`);
        (0, file_watcher_sockets_1.removeRegisteredFileWatcherSocket)(socket);
        (0, project_graph_listener_sockets_1.removeRegisteredProjectGraphListenerSocket)(socket);
    });
    socket.on('close', () => {
        numberOfOpenConnections -= 1;
        exports.openSockets.delete(socket);
        logger_1.serverLogger.log(`Closed a connection. Number of open connections: ${numberOfOpenConnections}`);
        (0, file_watcher_sockets_1.removeRegisteredFileWatcherSocket)(socket);
        (0, project_graph_listener_sockets_1.removeRegisteredProjectGraphListenerSocket)(socket);
    });
});
registerProcessTerminationListeners();
async function handleMessage(socket, data) {
    if (workspaceWatcherError) {
        await (0, shutdown_utils_1.respondWithErrorAndExit)(socket, `File watcher error in the workspace '${workspace_root_1.workspaceRoot}'.`, workspaceWatcherError);
    }
    (0, shutdown_utils_1.resetInactivityTimeout)(handleInactivityTimeout);
    const unparsedPayload = data;
    let payload;
    let mode = 'json';
    logger_1.serverLogger.log(`Received raw message of length ${unparsedPayload.length}`);
    try {
        // JSON Message
        if ((0, consume_messages_from_socket_1.isJsonMessage)(unparsedPayload)) {
            payload = JSON.parse(unparsedPayload);
        }
        else {
            // V8 Serialized Message
            payload = (0, v8_1.deserialize)(Buffer.from(unparsedPayload, 'binary'));
            mode = 'v8';
        }
    }
    catch (e) {
        await (0, shutdown_utils_1.respondWithErrorAndExit)(socket, `Invalid payload from the client`, new Error(`Unsupported payload sent to daemon server: ${unparsedPayload}`));
    }
    logger_1.serverLogger.log(`Received ${mode} message of type ${payload.type}`);
    if ((0, daemon_message_1.isDaemonMessage)(payload) && payload.env) {
        const envChanged = (0, daemon_environment_1.applyDaemonEnvFromClient)(payload.env);
        if (envChanged) {
            logger_1.serverLogger.log('Graph recompute necessary due to env variable refresh');
            forwardEnvToPluginWorkers(payload.env);
            (0, project_graph_incremental_recomputation_1.invalidateGraphCache)();
        }
    }
    if (payload.type === 'PING') {
        await handleResult(socket, 'PING', () => Promise.resolve({ response: true, description: 'ping' }), mode);
    }
    else if (payload.type === 'REQUEST_PROJECT_GRAPH') {
        await handleResult(socket, 'REQUEST_PROJECT_GRAPH', () => (0, handle_request_project_graph_1.handleRequestProjectGraph)(socket), mode);
    }
    else if (payload.type === 'HASH_TASKS') {
        await handleResult(socket, 'HASH_TASKS', () => (0, handle_hash_tasks_1.handleHashTasks)(payload), mode);
    }
    else if (payload.type === 'PROCESS_IN_BACKGROUND') {
        await handleResult(socket, 'PROCESS_IN_BACKGROUND', () => (0, handle_process_in_background_1.handleProcessInBackground)(payload), mode);
    }
    else if (payload.type === 'RECORD_OUTPUTS_HASH_BATCH') {
        await handleResult(socket, 'RECORD_OUTPUTS_HASH_BATCH', () => (0, handle_outputs_tracking_1.handleRecordOutputsHashBatch)(payload), mode);
    }
    else if (payload.type === 'OUTPUTS_HASHES_MATCH_BATCH') {
        await handleResult(socket, 'OUTPUTS_HASHES_MATCH_BATCH', () => (0, handle_outputs_tracking_1.handleOutputsHashesMatchBatch)(payload), mode);
    }
    else if (payload.type === 'REQUEST_SHUTDOWN') {
        await handleResult(socket, 'REQUEST_SHUTDOWN', () => (0, handle_request_shutdown_1.handleRequestShutdown)(server, numberOfOpenConnections), mode);
    }
    else if (payload.type === 'REGISTER_FILE_WATCHER') {
        file_watcher_sockets_1.registeredFileWatcherSockets.push({ socket, config: payload.config });
    }
    else if ((0, register_project_graph_listener_1.isRegisterProjectGraphListenerMessage)(payload)) {
        project_graph_listener_sockets_1.registeredProjectGraphListenerSockets.push(socket);
    }
    else if ((0, glob_1.isHandleGlobMessage)(payload)) {
        await handleResult(socket, glob_1.GLOB, () => (0, handle_glob_1.handleGlob)(payload.globs, payload.exclude), mode);
    }
    else if ((0, glob_1.isHandleMultiGlobMessage)(payload)) {
        await handleResult(socket, glob_1.MULTI_GLOB, () => (0, handle_glob_1.handleMultiGlob)(payload.globs, payload.exclude), mode);
    }
    else if ((0, get_nx_workspace_files_1.isHandleNxWorkspaceFilesMessage)(payload)) {
        await handleResult(socket, get_nx_workspace_files_1.GET_NX_WORKSPACE_FILES, () => (0, handle_nx_workspace_files_1.handleNxWorkspaceFiles)(payload.projectRootMap), mode);
    }
    else if ((0, get_files_in_directory_1.isHandleGetFilesInDirectoryMessage)(payload)) {
        await handleResult(socket, get_files_in_directory_1.GET_FILES_IN_DIRECTORY, () => (0, handle_get_files_in_directory_1.handleGetFilesInDirectory)(payload.dir), mode);
    }
    else if ((0, get_context_file_data_1.isHandleContextFileDataMessage)(payload)) {
        await handleResult(socket, get_context_file_data_1.GET_CONTEXT_FILE_DATA, () => (0, handle_context_file_data_1.handleContextFileData)(), mode);
    }
    else if ((0, hash_glob_1.isHandleHashGlobMessage)(payload)) {
        await handleResult(socket, hash_glob_1.HASH_GLOB, () => (0, handle_hash_glob_1.handleHashGlob)(payload.globs, payload.exclude), mode);
    }
    else if ((0, hash_glob_1.isHandleHashMultiGlobMessage)(payload)) {
        await handleResult(socket, hash_glob_1.HASH_GLOB, () => (0, handle_hash_glob_1.handleHashMultiGlob)(payload.globGroups), mode);
    }
    else if ((0, task_history_1.isHandleGetFlakyTasksMessage)(payload)) {
        await handleResult(socket, task_history_1.GET_FLAKY_TASKS, () => (0, handle_task_history_1.handleGetFlakyTasks)(payload.hashes), mode);
    }
    else if ((0, task_history_1.isHandleGetEstimatedTaskTimings)(payload)) {
        await handleResult(socket, task_history_1.GET_ESTIMATED_TASK_TIMINGS, () => (0, handle_task_history_1.handleGetEstimatedTaskTimings)(payload.targets), mode);
    }
    else if ((0, task_history_1.isHandleWriteTaskRunsToHistoryMessage)(payload)) {
        await handleResult(socket, task_history_1.RECORD_TASK_RUNS, () => (0, handle_task_history_1.handleRecordTaskRuns)(payload.taskRuns), mode);
    }
    else if ((0, force_shutdown_1.isHandleForceShutdownMessage)(payload)) {
        await handleResult(socket, 'FORCE_SHUTDOWN', () => (0, handle_force_shutdown_1.handleForceShutdown)(server), mode);
    }
    else if ((0, get_sync_generator_changes_1.isHandleGetSyncGeneratorChangesMessage)(payload)) {
        await handleResult(socket, get_sync_generator_changes_1.GET_SYNC_GENERATOR_CHANGES, () => (0, handle_get_sync_generator_changes_1.handleGetSyncGeneratorChanges)(payload.generators), mode);
    }
    else if ((0, flush_sync_generator_changes_to_disk_1.isHandleFlushSyncGeneratorChangesToDiskMessage)(payload)) {
        await handleResult(socket, flush_sync_generator_changes_to_disk_1.FLUSH_SYNC_GENERATOR_CHANGES_TO_DISK, () => (0, handle_flush_sync_generator_changes_to_disk_1.handleFlushSyncGeneratorChangesToDisk)(payload.generators), mode);
    }
    else if ((0, get_registered_sync_generators_1.isHandleGetRegisteredSyncGeneratorsMessage)(payload)) {
        await handleResult(socket, get_registered_sync_generators_1.GET_REGISTERED_SYNC_GENERATORS, () => (0, handle_get_registered_sync_generators_1.handleGetRegisteredSyncGenerators)(), mode);
    }
    else if ((0, update_workspace_context_1.isHandleUpdateWorkspaceContextMessage)(payload)) {
        await handleResult(socket, update_workspace_context_1.UPDATE_WORKSPACE_CONTEXT, () => (0, handle_update_workspace_context_1.handleUpdateWorkspaceContext)(payload.createdFiles, payload.updatedFiles, payload.deletedFiles), mode);
    }
    else if ((0, run_tasks_execution_hooks_1.isHandlePreTasksExecutionMessage)(payload)) {
        await handleResult(socket, run_tasks_execution_hooks_1.PRE_TASKS_EXECUTION, () => (0, handle_tasks_execution_hooks_1.handleRunPreTasksExecution)(payload.context), mode);
    }
    else if ((0, run_tasks_execution_hooks_1.isHandlePostTasksExecutionMessage)(payload)) {
        await handleResult(socket, run_tasks_execution_hooks_1.POST_TASKS_EXECUTION, () => (0, handle_tasks_execution_hooks_1.handleRunPostTasksExecution)(payload.context), mode);
    }
    else if ((0, nx_console_1.isHandleGetNxConsoleStatusMessage)(payload)) {
        await handleResult(socket, nx_console_1.GET_NX_CONSOLE_STATUS, () => (0, handle_nx_console_1.handleGetNxConsoleStatus)(), mode);
    }
    else if ((0, nx_console_1.isHandleSetNxConsolePreferenceAndInstallMessage)(payload)) {
        await handleResult(socket, nx_console_1.SET_NX_CONSOLE_PREFERENCE_AND_INSTALL, () => (0, handle_nx_console_1.handleSetNxConsolePreferenceAndInstall)(payload.preference), mode);
    }
    else if ((0, configure_ai_agents_1.isHandleGetConfigureAiAgentsStatusMessage)(payload)) {
        await handleResult(socket, configure_ai_agents_1.GET_CONFIGURE_AI_AGENTS_STATUS, () => (0, handle_configure_ai_agents_1.handleGetConfigureAiAgentsStatus)(), mode);
    }
    else if ((0, configure_ai_agents_1.isHandleResetConfigureAiAgentsStatusMessage)(payload)) {
        await handleResult(socket, configure_ai_agents_1.RESET_CONFIGURE_AI_AGENTS_STATUS, () => (0, handle_configure_ai_agents_1.handleResetConfigureAiAgentsStatus)(), mode);
    }
    else {
        await (0, shutdown_utils_1.respondWithErrorAndExit)(socket, `Invalid payload from the client`, new Error(`Unsupported payload sent to daemon server: ${unparsedPayload}`));
    }
}
async function handleResult(socket, type, hrFn, mode) {
    let hr;
    const startMark = new Date();
    try {
        hr = await hrFn();
    }
    catch (error) {
        hr = { description: `[${type}]`, error };
    }
    const doneHandlingMark = new Date();
    if (hr.error) {
        await (0, shutdown_utils_1.respondWithErrorAndExit)(socket, hr.description, hr.error);
    }
    else {
        logger_1.serverLogger.log(`Serializing response for ${type} message in ${mode} mode`);
        const response = typeof hr.response === 'string'
            ? hr.response
            : serializeUnserializedResult(hr.response, mode);
        logger_1.serverLogger.log(`Responding to ${type} message`);
        await (0, shutdown_utils_1.respondToClient)(socket, response, hr.description);
    }
    const endMark = new Date();
    logger_1.serverLogger.log(`Handled ${mode} message ${type}. Handling time: ${doneHandlingMark.getTime() - startMark.getTime()}. Response time: ${endMark.getTime() - doneHandlingMark.getTime()}.`);
}
function handleInactivityTimeout() {
    if ((0, file_watcher_sockets_1.hasRegisteredFileWatcherSockets)() ||
        (0, project_graph_listener_sockets_1.hasRegisteredProjectGraphListenerSockets)()) {
        logger_1.serverLogger.log(`There are open file watchers or project graph listeners. Resetting inactivity timer.`);
        (0, shutdown_utils_1.resetInactivityTimeout)(handleInactivityTimeout);
    }
    else {
        (0, shutdown_utils_1.handleServerProcessTermination)({
            server,
            reason: `${shutdown_utils_1.SERVER_INACTIVITY_TIMEOUT_MS}ms of inactivity`,
            sockets: exports.openSockets,
        });
    }
}
function registerProcessTerminationListeners() {
    process
        .on('SIGINT', () => (0, shutdown_utils_1.handleServerProcessTermination)({
        server,
        reason: 'received process SIGINT',
        sockets: exports.openSockets,
    }))
        .on('SIGTERM', () => (0, shutdown_utils_1.handleServerProcessTermination)({
        server,
        reason: 'received process SIGTERM',
        sockets: exports.openSockets,
    }))
        .on('SIGHUP', () => (0, shutdown_utils_1.handleServerProcessTermination)({
        server,
        reason: 'received process SIGHUP',
        sockets: exports.openSockets,
    }));
}
let existingLockHash;
function daemonIsOutdated() {
    if ((0, is_nx_version_mismatch_1.isNxVersionMismatch)()) {
        return 'NX_VERSION_CHANGED';
    }
    else if (lockFileHashChanged()) {
        return 'LOCK_FILES_CHANGED';
    }
    return null;
}
function lockFileHashChanged() {
    const lockFiles = [
        (0, path_1.join)(workspace_root_1.workspaceRoot, 'package-lock.json'),
        (0, path_1.join)(workspace_root_1.workspaceRoot, 'yarn.lock'),
        (0, path_1.join)(workspace_root_1.workspaceRoot, 'pnpm-lock.yaml'),
        (0, path_1.join)(workspace_root_1.workspaceRoot, 'bun.lockb'),
        (0, path_1.join)(workspace_root_1.workspaceRoot, 'bun.lock'),
    ];
    const existingFiles = lockFiles.filter((file) => (0, fs_1.existsSync)(file));
    const lockHashes = existingFiles.map((file) => (0, native_1.hashFile)(file));
    const newHash = (0, file_hasher_1.hashArray)(lockHashes);
    if (existingLockHash && newHash != existingLockHash) {
        logger_1.serverLogger.log(`[Server] lock file hash changed! old=${existingLockHash}, new=${newHash}`);
        existingLockHash = newHash;
        return true;
    }
    else {
        existingLockHash = newHash;
        return false;
    }
}
/**
 * When applicable files in the workspaces are changed (created, updated, deleted),
 * we need to recompute the cached serialized project graph so that it is readily
 * available for the next client request to the server.
 */
const handleWorkspaceChanges = async (err, changeEvents) => {
    if (workspaceWatcherError) {
        logger_1.serverLogger.watcherLog('Skipping handleWorkspaceChanges because of a previously recorded watcher error.');
        return;
    }
    try {
        (0, shutdown_utils_1.resetInactivityTimeout)(handleInactivityTimeout);
        if (err) {
            let error = typeof err === 'string' ? new Error(err) : err;
            logger_1.serverLogger.watcherLog('Unexpected workspace watcher error', error.message);
            console.error(error);
            workspaceWatcherError = error;
            return;
        }
        logger_1.serverLogger.watcherLog((0, watcher_1.convertChangeEventsToLogMessage)(changeEvents));
        (0, route_workspace_changes_1.routeWorkspaceChanges)(changeEvents);
    }
    catch (err) {
        logger_1.serverLogger.watcherLog(`Unexpected workspace error`, err.message);
        console.error(err);
        workspaceWatcherError = err;
    }
};
const handleOutputsChanges = async (err, changeEvents) => {
    try {
        if (err || !changeEvents || !changeEvents.length) {
            let error = typeof err === 'string' ? new Error(err) : err;
            logger_1.serverLogger.watcherLog('Unexpected outputs watcher error', error.message);
            console.error(error);
            outputsWatcherError = error;
            (0, outputs_tracking_1.disableOutputsTracking)();
            return;
        }
        if (outputsWatcherError) {
            return;
        }
        logger_1.serverLogger.watcherLog('Processing file changes in outputs');
        (0, outputs_tracking_1.processFileChangesInOutputs)(changeEvents);
    }
    catch (err) {
        logger_1.serverLogger.watcherLog(`Unexpected outputs watcher error`, err.message);
        console.error(err);
        outputsWatcherError = err;
        (0, outputs_tracking_1.disableOutputsTracking)();
    }
};
async function startServer() {
    (0, workspace_context_1.setupWorkspaceContext)(workspace_root_1.workspaceRoot);
    // Initialize analytics for daemon process
    await (0, analytics_1.startAnalytics)();
    const socketPath = (0, socket_utils_1.getFullOsSocketPath)();
    // Log daemon startup information for debugging
    logger_1.serverLogger.log(`New daemon starting from: ${__filename}`);
    logger_1.serverLogger.log(`New daemon __dirname: ${__dirname}`);
    logger_1.serverLogger.log(`New daemon nxVersion: ${versions_1.nxVersion}`);
    logger_1.serverLogger.log(`New daemon getInstalledNxVersion(): ${(0, installed_nx_version_1.getInstalledNxVersion)()}`);
    // Persist metadata about the background process so that it can be cleaned up later if needed
    await (0, cache_1.writeDaemonJsonProcessCache)({
        processId: process.pid,
        socketPath,
        nxVersion: versions_1.nxVersion,
    });
    // See notes in socket-command-line-utils.ts on OS differences regarding clean up of existings connections.
    if (!socket_utils_1.isWindows) {
        (0, socket_utils_1.killSocketOrPath)();
    }
    logger_1.serverLogger.log(`[Server] Starting outdated check interval (20ms)`);
    setInterval(() => {
        if ((0, cache_1.getDaemonProcessIdSync)() !== process.pid) {
            return (0, shutdown_utils_1.handleServerProcessTermination)({
                server,
                reason: 'this process is no longer the current daemon (native)',
                sockets: exports.openSockets,
            });
        }
        const outdated = daemonIsOutdated();
        if (outdated) {
            logger_1.serverLogger.log(`[Server] Daemon outdated: ${outdated}`);
            if (outdated === 'LOCK_FILES_CHANGED') {
                // Lock file changes - restart daemon, clients will reconnect
                logger_1.serverLogger.log('[Server] Restarting daemon...');
                (0, shutdown_utils_1.handleServerProcessTerminationWithRestart)({
                    server,
                    reason: outdated,
                    sockets: exports.openSockets,
                });
            }
            else {
                // Version changes or other reasons - just shut down, don't restart
                logger_1.serverLogger.log('[Server] Shutting down daemon (no restart)...');
                (0, shutdown_utils_1.handleServerProcessTermination)({
                    server,
                    reason: outdated,
                    sockets: exports.openSockets,
                });
            }
        }
    }, 20).unref();
    return new Promise(async (resolve, reject) => {
        try {
            server.listen(socketPath, async () => {
                try {
                    logger_1.serverLogger.log(`Started listening on: ${socketPath}`);
                    // this triggers the storage of the lock file hash
                    daemonIsOutdated();
                    if (!(0, shutdown_utils_1.getWatcherInstance)()) {
                        (0, shutdown_utils_1.storeWatcherInstance)(await (0, watcher_1.watchWorkspace)(server, handleWorkspaceChanges));
                        logger_1.serverLogger.watcherLog(`Subscribed to changes within: ${workspace_root_1.workspaceRoot} (native)`);
                    }
                    if (!(0, shutdown_utils_1.getOutputWatcherInstance)()) {
                        (0, shutdown_utils_1.storeOutputWatcherInstance)(await (0, watcher_1.watchOutputFiles)(server, handleOutputsChanges));
                    }
                    // listen for project graph recomputation events to collect and schedule sync generators
                    (0, project_graph_incremental_recomputation_1.registerProjectGraphRecomputationListener)(sync_generators_1.collectAndScheduleSyncGenerators);
                    // register file change listener to invalidate sync generator cache
                    (0, file_change_events_1.registerFileChangeListener)(sync_generators_1.clearSyncGeneratorsCache);
                    // trigger an initial project graph recomputation
                    (0, project_graph_incremental_recomputation_1.scheduleProjectGraphRecomputation)([], [], []);
                    // Kick off Nx Console check in background to prime the cache
                    (0, handle_nx_console_1.handleGetNxConsoleStatus)().catch(() => {
                        // Ignore errors, this is a background operation
                    });
                    // Kick off AI agents outdated check in background to prime the cache
                    (0, handle_configure_ai_agents_1.handleGetConfigureAiAgentsStatus)().catch(() => {
                        // Ignore errors, this is a background operation
                    });
                    return resolve(server);
                }
                catch (err) {
                    await handleWorkspaceChanges(err, []);
                }
            });
        }
        catch (err) {
            reject(err);
        }
    });
}
function forwardEnvToPluginWorkers(env) {
    (0, get_plugins_1.getPlugins)((0, nx_json_1.readNxJson)(workspace_root_1.workspaceRoot))
        .then((plugins) => {
        for (const plugin of plugins) {
            plugin.setWorkerEnv?.(env)?.catch((e) => {
                logger_1.serverLogger.log(`Failed to forward env to plugin worker "${plugin.name}": ${e.message}`);
            });
        }
    })
        .catch(() => {
        // Plugins may not be loaded yet — env will be picked up on next load
    });
}
function serializeUnserializedResult(response, mode) {
    if (mode === 'json') {
        return JSON.stringify(response);
    }
    else {
        return (0, v8_1.serialize)(response).toString('binary');
    }
}
